### PR TITLE
[GISel][RISCV] Legalize `G_FREM` to use fmod

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch adds support for `G_FREM` on RISCV.

Unfortunately, I cannot provide vector tests for `scalarize` and half/bfloat16 tests for `minScalar`.

### CODE REVIEW

**Review by topperc — APPROVED** [review]
LGTM<

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 1]
```diff
@@ -372,6 +372,11 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
                                G_FABS, G_FSQRT, G_FMAXNUM, G_FMINNUM})
       .legalIf(typeIsScalarFPArith(0, ST));
 
+  getActionDefinitionsBuilder(G_FREM)
+      .libcallFor({s32, s64})
+      .minScalar(0, s32)
+      .scalarize(0);
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:378]: Are there tests for the `scalarize` case? I don't think we should scalarize if we don't have test covereage for it. I would prefer legalizing to fail.

**Review by michaelmaitland — CHANGES_REQUESTED** [review]

**Review by arsenm — COMMENTED** [review]
It should be possible to test the scalarize case

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 1 above)
Comment (dtcxzyw) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:378]: I will leave it as a todo.

**Comment by arsenm** [comment]
> > It should be possible to test the scalarize case
> 
> I tried the following 4 variants, but no one works :(

This is the case where MIR tests are useful  because you don't have to bring up all the other parts

File: llvm/test/CodeGen/RISCV/GlobalISel/libcalls.ll [hunk 2]
```diff
@@ -0,0 +1,51 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=riscv32 -global-isel -verify-machineinstrs < %s \
+; RUN:   | FileCheck %s --check-prefix=RV32
+; RUN: llc -mtriple=riscv64 -global-isel -verify-machineinstrs < %s \
+; RUN:   | FileCheck %s --check-prefix=RV64
+
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/libcalls.ll:6]: We don't support callconv/G_FPTRUNC/G_FPEXT for half types. I plan to implement zfh support next month.

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/CodeGen/RISCV/GlobalISel/libcalls.ll:6]: You can work around that with i16 and bitcast to half

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/CodeGen/RISCV/GlobalISel/libcalls.ll:6]: Also again, MIR tests avoid needing to handle everything

**Comment by dtcxzyw** [comment]
> > > It should be possible to test the scalarize case
> > 
> > 
> > I tried the following 4 variants, but no one works :(
> 
> This is the case where MIR tests are useful because you don't have to bring up all the other parts

These MIR tests don't work:
```
---
name:            frem_v2f32
body:             |
  bb.0.entry:
    ; unable to legalize instruction: %3:_(s32), %4:_(s32) = G_UNMERGE_VALUES %0:_(<2 x s32>)
    %0:_(<2 x s32>) = COPY $v8
    %1:_(<2 x s32>) = COPY $v9
    %2:_(<2 x s32>) = G_FREM %0, %1
    $v8 = COPY %2(<2 x s32>)
    PseudoRET implicit $v8

...
---
name:            frem_nxv2f32
body:             |
  bb.0.entry:
    ; Invalid size request on a scalable vector.
    %0:_(<vscale x 2 x s32>) = COPY $v8
    %1:_(<vscale x 2 x s32>) = COPY $v9
    %2:_(<vscale x 2 x s32>) = G_FREM %0, %1
    $v8 = COPY %2(<vscale x 2 x s32>)
    PseudoRET implicit $v8

...
```

File: llvm/test/CodeGen/RISCV/GlobalISel/libcalls.ll (hunk 2 above)
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/libcalls.ll:6]: > You can work around that with i16 and bitcast to half

I mean widening half to float will introduce G_FPTRUNC and G_FPEXT.
https://github.com/llvm/llvm-project/blob/32c3561d44aa792ef08d72b5a4c342c9965bc4c2/llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp#L2891-L2917

**Comment by arsenm** [comment]
> > > > It should be possible to test the scalarize case
> > > 
> > > 
> > > I tried the following 4 variants, but no one works :(
> > 
> > 
> > This is the case where MIR tests are useful because you don't have to bring up all the other parts
> 
> These MIR tests don't work:
> 
> ```
> ---
> name:            frem_v2f32
> body:             |
>   bb.0.entry:
>     ; unable to legalize instruction: %3:_(s32), %4:_(s32) = G_UNMERGE_VALUES %0:_(<2 x s32>)

If you use -global-isel-abort=2, you can check the partially legalized output

**Comment by dtcxzyw** [comment]
> > > > > It should be possible to test the scalarize case
> > > > 
> > > > 
> > > > I tried the following 4 variants, but no one works :(
> > > 
> > > 
> > > This is the case where MIR tests are useful because you don't have to bring up all the other parts
> > 
> > 
> > These MIR tests don't work:
> > ```
> > ---
> > name:            frem_v2f32
> > body:             |
> >   bb.0.entry:
> >     ; unable to legalize instruction: %3:_(s32), %4:_(s32) = G_UNMERGE_VALUES %0:_(<2 x s32>)
> > ```
> 
> If you use -global-isel-abort=0 or 2, you can check the partially legalized output

Done. Thank you!

**Review by arsenm — APPROVED** [review]

**Pull request merged**